### PR TITLE
migrating(eks): add blog post of migrating node groups with zero downtime

### PR DESCRIPTION
**Pull request by metral**
Blog post for EKS migrate-nodegroups example:

EKS examples PR: https://github.com/pulumi/pulumi-eks/pull/195
Tutorial PR: https://github.com/pulumi/docs/pull/1327

### CODE REVIEW

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+If you'd like to chat with our team, or get hands-on assistance with
+migrating your existing configuration code (including ksonnet programs)
+to Pulumi, please don't hesitate to [drop us a line]({{< ref "/contact" >}}).
+
+[eks-amis]: https://docs.aws.amazon.com/eks/latest/userguide/eks-optimized-ami.html
+[ingress-nginx]: https://github.com/kubernetes/ingress-nginx
+[echoserver]: https://github.com/kubernetes-retired/contrib/blob/master/ingress/echoheaders/echo-app.yaml
+[pod-lifecycle]: https://kubernetes.io/docs/concepts/workloads/pods/pod/#termination-of-pods
+[eks-nodegroup-tutorial]: ({{< relref "/docs/reference/tutorials/kubernetes/tutorial-eks" >}})
+[crosswalk-aws]: ({{< relref "/docs/reference/crosswalk/aws" >}}).
```
Comment (metral) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]: Making a note to fix `relref` hugo URL once the tutorial is live

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
+---
+title: "Day 2 Kubernetes - Migrating EKS Node Groups with Zero Downtime"
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
title: "Day 2 Kubernetes: Migrating EKS Node Groups with Zero Downtime"
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+meta_image: "eks-migrate-nodegroups.svg"
+---
+
+Managed Kubernetes clusters greatly reduce the overhead required in
+administering Kubernetes. However, the cluster is only one of the
+components under management, as app lifecycles are self-driven tasks
+that vary by workloads.
+
+In Kubernetes, node groups are a useful mechanism of creating pools of resources that can enforce
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
In Kubernetes, node groups are a useful mechanism for creating pools of resources that can enforce
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
+---
+title: "Day 2 Kubernetes - Migrating EKS Node Groups with Zero Downtime"
+authors: ["mike-metral"]
+tags: ["Kubernetes","EKS", "Migrations", "Node Groups"]
+date: "2019-07-18"
+
+meta_image: "eks-migrate-nodegroups.svg"
+---
+
+Managed Kubernetes clusters greatly reduce the overhead required in
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]: This somehow reads a little incomplete. Do you mean "administering Kubernetes deployments"?

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+administering Kubernetes. However, the cluster is only one of the
+components under management, as app lifecycles are self-driven tasks
+that vary by workloads.
+
+In Kubernetes, node groups are a useful mechanism of creating pools of resources that can enforce
+scheduling requirements. They also provide a utility for shifting
+workloads around during cluster management and updates.
+
+In this post, we'll see how to use Pulumi for Day 2 Kubernetes administration.
+We'll spin up a new EKS cluster with two node groups, and a given workload.
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
We'll spin up a new EKS cluster with two node groups and a given workload.
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+Then we'll add an additional, updated node group, and migrate the workload
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]: I'd rephrase "add an additional" to "add one more" but I'm not sure what you mean by "updated". Will "add one more updated node group" work as a substitution here?

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+over to it with zero downtime using code and `kubectl`.
+
+<!--more-->
+
+[View the full tutorial and code.][eks-nodegroup-tutorial]
+
+<p align="center"><img src="eks-migrate-nodegroups.svg" width="650"/></p>
+
+## Create an EKS Cluster & Deploy Workload
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
## Create an EKS cluster and deploy the workload
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+The inital update we'll perform will configure and launch an EKS cluster using
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
For our initial update, we will configure and launch an EKS cluster using
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+`v1.13` of Kubernetes, along with the cluster's infrastructure dependencies
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
`v1.13` of Kubernetes, with the cluster's infrastructure dependencies 
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+(VPC and IAM) using [Crosswalk for AWS][crosswalk-aws]. We'll also
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
(such as VPC and IAM) defined using [Crosswalk for AWS][crosswalk-aws]. We'll also
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+create and attach the following node groups to the cluster:
+
+* A standard `t2.medium` worker node group using the recent `v1.13.7` worker [AMI][eks-amis]. For use by general purpose workloads such as the [`EchoServer`][echoserver], a simple app that echo's client request headers.
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
* A standard `t2.medium` worker node group using the recent `v1.13.7` worker [AMI][eks-amis], for general purpose workloads such as the [`EchoServer`][echoserver] &mdash; a simple app that echo's client request headers.
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+* A 2xlarge `t3.2xlarge` worker node group using an older `v1.12.7` worker [AMI][eks-amis]. For use use by larger, intensive workloads such as the [NGINX Ingress Controller][ingress-nginx].
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
* A 2xlarge `t3.2xlarge` worker node group using an older `v1.12.7` worker [AMI][eks-amis], for use by larger, intensive workloads such as the [NGINX Ingress Controller][ingress-nginx].
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+
+With the cluster successfully created and the node groups available, Pulumi
+will deploy the workload: the [`EchoServer`][echoserver], and the [NGINX Ingress Controller][ingress-nginx] that will manage it's ingress.
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
will deploy the workload: the [`EchoServer`][echoserver], and the [NGINX Ingress Controller][ingress-nginx] that will manage its ingress.
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+Once the workload is deployed, we can validate it is up and running by accessing
+the `echoserver` behind the NGINX endpoint using `curl`:
+
+<img style="max-width:none; height: 350px; width: 850px;" src="k8s-cluster-workload.gif"/>
+
+> **Note**: You can open the <a href="./k8s-cluster-workload.gif" target="_blank">GIF</a> for a maximized view.
+
+## The Great Migration
+
+After initial deployment, we decide to update the node group used by NGINX.
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
After our initial deployment, let's say we decide to update the node group used by NGINX.
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+
+We'll move NGINX from the `2xlarge` node group over to a new, `4xlarge` worker
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
To do so, we'll move NGINX from the `2xlarge` node group over to a new, `4xlarge` worker
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+node group that differs in: an updated AMI, instance type, and desired instance count.
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
node group with an updated AMI, a different instance type, and a different desired instance count.
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+
+As we migrate NGINX over to the `4xlarge` and decommission the `2xlarge` node
+group in the next steps, we'll also actively load test the endpoint of the
+`echoserver` to ensure that we are not losing requests through out the migration.
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
`echoserver` to ensure that we are not losing requests throughout the migration.
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+
+### Step 1: Create the new `4xlarge` Node Group
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
### Step 1: Create the new `4xlarge` node group.
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+
+Create the new, `4xlarge` node group by defining it and running an update.
+
+<p align="center"><img src="ng-4xlarge.svg" width="650"/></p>
+
+### Step 2: Migrate NGINX to the `4xlarge` Node Group
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
### Step 2: Migrate NGINX to the `4xlarge` node group.
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 20]
```diff
@@ -0,0 +1,170 @@
...
+With the `4xlarge` node group created, we'll migrate the NGINX service away
+from the `2xlarge` node group over to the `4xlarge` node group by changing
+its node selector scheduling terms.
+
+This change updates the NGINX Deployment spec to require the use of
+`c5.4xlarge` nodes during scheduling, and forces a rolling update over to the
+`4xlarge` node group.
+
+NGINX is able to successfully migrate across node groups because it is
+configured with HA settings, spread-type scheduling predicates,
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]: spread-type scheduling predicates - Is there a good link for this?

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 21]
```diff
@@ -0,0 +1,170 @@
...
+and can gracefully terminate within the Kubernetes [Pod lifecycle][pod-lifecycle].
+
+<p align="center"><img src="target-ng-4xlarge.svg" width="650"/></p>
+
+### Step 3: Decomission the `2xlarge` Node Group
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
### Step 3: Decommission the `2xlarge` node group.

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 22]
```diff
@@ -0,0 +1,170 @@
...
+
+Once migration of NGINX has completed onto the `4xlarge` node group, we can
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
Once NGINX has been completely migrated to the `4xlarge` node group, we can
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 23]
```diff
@@ -0,0 +1,170 @@
...
+commence the decommissioning of the original `2xlarge` node group no
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
begin decommissioning the original `2xlarge` node group which is no longer in use.
```

I would just end with that instead of adding "from Kubernetes and AWS"

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 24]
```diff
@@ -0,0 +1,170 @@
...
+longer in use from Kubernetes and AWS.
+
+Decommissioning the node group means that we'll:
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
Decommissioning the node group involves:
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 25]
```diff
@@ -0,0 +1,170 @@
...
+
+  * Drain the Kubernetes nodes.
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
  * Draining the Kubernetes nodes.
```

Same for all the list items.

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 26]
```diff
@@ -0,0 +1,170 @@
...
+  * Delete the Kubernetes nodes from the APIServer.
+  * Scale down the Auto Scaling Group to `0`.
+  * Delete the node group.
+
+Setup `kubectl` by using the `kubeconfig` from the stack output.
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
Set up `kubectl` by using the `kubeconfig` from the stack output.
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 27]
```diff
@@ -0,0 +1,170 @@
...
+
+To drain the `2xlarge` node group, we'll use `kubectl drain`:
+
+```bash
+for node in $(kubectl get nodes -l beta.kubernetes.io/instance-type=t3.2xlarge -o=name); do
+    kubectl drain --force --ignore-daemonsets --delete-local-data --grace-period=10 "$node";
+done
+```
+
+After draining has completed, to delete the nodes from the APIServer use `kubectl delete node`:
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
After draining has completed, use `kubectl delete node` to delete the nodes from the APIServer.
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 28]
```diff
@@ -0,0 +1,170 @@
...
+    kubectl delete "$node";
+done
+```
+
+Scale down the node group Auto Scaling Group completely by setting
+the `desiredCapacity: 0` and running an update:
+
+<p align="center"><img src="scale-down-ng-2xlarge.svg" width="650"/></p>
+
+Once the Auto Scaling Group has scaled down, we can delete the node group from
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
Once the Auto Scaling group has scaled down, we can delete the node group from
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 29]
```diff
@@ -0,0 +1,170 @@
...
+AWS and the Pulumi program:
+
+<p align="center"><img src="remove-ng-2xlarge.svg" width="650"/></p>
+
+## Summary
+
+In this post we stood up an EKS cluster with a couple of node groups, and an
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
In this post, we stood up an EKS cluster with a couple of node groups, and an
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 30]
```diff
@@ -0,0 +1,170 @@
...
+`echoserver` and NGINX workload. We then created a new, updated node group and
+migrated NGINX over to it.
+
+We acheived this node group migration with zero downtime to our apps during
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
We achieved this node group migration with zero downtime to our apps during
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 31]
```diff
@@ -0,0 +1,170 @@
...
+load testing and decommissioning of the original node group.
+
+<img style="max-width:none; height: 350px; width: 850px;" src="eks-migration.gif"/>
+
+> **Note**: You can open the <a href="./eks-migration.gif" target="_blank">GIF</a> for a maximized view.
+
+## Learn More
+
+If you'd like to learn about Pulumi and how to manage your
+infrastructure and Kubernetes through code, [click here to get started today]({{< ref "/docs/quickstart" >}}). Pulumi is open source and free to
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
infrastructure and Kubernetes through code, [get started today]({{< ref "/docs/quickstart" >}}). Pulumi is open source and free to
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 32]
```diff
@@ -0,0 +1,170 @@
...
+use.
+
+For a follow-up example on how to further use Pulumi to create Kubernetes
```
Comment (bermudezmt) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
For further examples on how to use Pulumi to create Kubernetes
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md (hunk 4 above)
Comment (metral) [reply content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]: I was referring to Managed Kubernetes offerings like EKS. I've changed clusters -> offerings to make this more straight-forward.

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md (hunk 6 above)
Comment (metral) [reply content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]: Rephrased to: "Then we'll add one more node group with updated configuration..."

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md (hunk 20 above)
Comment (metral) [reply content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]: Linked to the HA + refs section in the tutorial:
https://www.pulumi.com/docs/reference/tutorials/kubernetes/tutorial-eks-migrate-nodegroups/#kubernetes-workloads-high-availability-ha-rolling-updates

**Comment by metral** [comment]
Feedback has been addressed. PTAL @bermudezmt

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 33]
```diff
@@ -0,0 +1,172 @@
+---
+title: "Day 2 Kubernetes - Migrating EKS Node Groups with Zero Downtime"
+authors: ["mike-metral"]
+tags: ["Kubernetes","EKS", "Migrations", "Node Groups"]
+date: "2019-07-22"
+
+meta_image: "eks-migrate-nodegroups.png"
+meta_description: "Using Pulumi, we'll walk through Day 2 Kubernetes cluster administration by migrating a workload across EKS node groups with zero downtime."
```
Comment (cnunciato) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]: This field is actually optional (and unused in the new blog), but if you do want to specify an explicit `<meta name="description" content="[...]">`, the property should be named `meta_desc`, not `meta_description`. When omitted, we just use the `.Summary` property supplied by Hugo (which is derived from the `<!--more-->` comment you've provided in your post). In general, if you've written an intro paragraph suitable for a home-page and/or social-media preview, you needn't specify anything for this.

Comment (metral) [reply content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]: Thanks for the tip! I've removed `meta_description` in favor of the `.Summary` which is better detailed.

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 34]
```diff
@@ -0,0 +1,171 @@
...
+## Summary
+
+In this post, we stood up an EKS cluster with a couple of node groups, and an
+`echoserver` and NGINX workload. We then created a new, updated node group and
+migrated NGINX over to it.
+
+We achieved this node group migration with zero downtime to our apps during
+load testing and decommissioning of the original node group.
+
+<img style="max-width:none; height: 350px; width: 850px;" src="eks-migration.gif"/>
```
Comment (cnunciato) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]: These can just be regular images, as they'll be sized to scale within the container. (We generally shouldn't need to specify explicit sizes on images, but where we do, we have some CSS utility classes for that &mdash; hit me up and I can point you to some examples.)

```suggestion
![](eks-migration.gif)
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 35]
```diff
@@ -0,0 +1,171 @@
...
+In this post, we'll see how to use Pulumi for Day 2 Kubernetes administration.
+We'll spin up a new EKS cluster with two node groups and a given workload.
+Then we'll add one more node group with an updated configuration, and migrate the workload
+over to it with zero downtime using code and `kubectl`.
+
+<!--more-->
+
+[View the full tutorial and code.][eks-nodegroup-tutorial]
+
+<p align="center"><img src="eks-migrate-nodegroups.svg" width="650"/></p>
```
Comment (cnunciato) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
![](eks-migrate-nodegroups.svg)
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 36]
```diff
@@ -0,0 +1,171 @@
...
+
+With the cluster successfully created and the node groups available, Pulumi
+will deploy the workload: the [`EchoServer`][echoserver], and the [NGINX Ingress Controller][ingress-nginx] that will manage its ingress.
+The `echoserver` will land on the Standard node group, and NGINX is set
+to specifically target the `2xlarge` node group.
+
+Once the workload is deployed, we can validate it is up and running by accessing
+the `echoserver` behind the NGINX endpoint using `curl`:
+
+<img style="max-width:none; height: 350px; width: 850px;" src="k8s-cluster-workload.gif"/>
```
Comment (cnunciato) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
![](k8s-cluster-workload.gif)
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 37]
```diff
@@ -0,0 +1,171 @@
...
+
+As we migrate NGINX over to the `4xlarge` and decommission the `2xlarge` node
+group in the next steps, we'll also actively load test the endpoint of the
+`echoserver` to ensure that we are not losing requests throughout the migration.
+
+### Step 1: Create the new `4xlarge` node group.
+
+Create the new, `4xlarge` node group by defining it and running an update.
+
+<p align="center"><img src="ng-4xlarge.svg" width="650"/></p>
```
Comment (cnunciato) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
![](ng-4xlarge.svg)
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 38]
```diff
@@ -0,0 +1,171 @@
...
+
+This change updates the NGINX Deployment spec to require the use of
+`c5.4xlarge` nodes during scheduling, and forces a rolling update over to the
+`4xlarge` node group.
+
+NGINX is able to successfully migrate across node groups because it is
+configured with HA settings, [spread-type scheduling predicates][tutorial-ha-refs],
+and can gracefully terminate within the Kubernetes [Pod lifecycle][pod-lifecycle].
+
+<p align="center"><img src="target-ng-4xlarge.svg" width="650"/></p>
```
Comment (cnunciato) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
![](target-ng-4xlarge.svg)
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 39]
```diff
@@ -0,0 +1,171 @@
...
+```bash
+for node in $(kubectl get nodes -l beta.kubernetes.io/instance-type=t3.2xlarge -o=name); do
+    kubectl delete "$node";
+done
+```
+
+Scale down the node group Auto Scaling Group completely by setting
+the `desiredCapacity: 0` and running an update:
+
+<p align="center"><img src="scale-down-ng-2xlarge.svg" width="650"/></p>
```
Comment (cnunciato) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
![](scale-down-ng-2xlarge.svg)
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 40]
```diff
@@ -0,0 +1,171 @@
...
+
+Once the Auto Scaling Group has scaled down, we can delete the node group from
+AWS and the Pulumi program:
+
+<p align="center"><img src="remove-ng-2xlarge.svg" width="650"/></p>
```
Comment (cnunciato) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
![](remove-ng-2xlarge.svg)
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 41]
```diff
@@ -0,0 +1,171 @@
...
+configured with HA settings, [spread-type scheduling predicates][tutorial-ha-refs],
+and can gracefully terminate within the Kubernetes [Pod lifecycle][pod-lifecycle].
+
+<p align="center"><img src="target-ng-4xlarge.svg" width="650"/></p>
+
+### Step 3: Decommission the `2xlarge` node group.
+
+Once NGINX has been completely migrated to the `4xlarge` node group, we can
+begin decommissioning the original `2xlarge` node group which is no longer in use.
+longer in use from Kubernetes and AWS.
```
Comment (lblackstone) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 42]
```diff
@@ -0,0 +1,171 @@
...
+
+[View the full tutorial and code.][eks-nodegroup-tutorial]
+
+<p align="center"><img src="eks-migrate-nodegroups.svg" width="650"/></p>
+
+## Create an EKS cluster and Deploy the Workload
+
+For our initial update, we will configure and launch an EKS cluster using
+`v1.13` of Kubernetes, with the cluster's infrastructure dependencies 
+(such as VPC and IAM) defined using [Crosswalk for AWS][crosswalk-aws]. We'll also
```
[the remark this answers is not in the document]
Comment (metral) [reply content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]: Done

**Comment by metral** [comment]
Feedback has been addressed. PTAL
@cnunciato @bermudezmt @lblackstone @aydrian

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 43]
```diff
@@ -0,0 +1,174 @@
+---
+title: "Day 2 Kubernetes: Migrating EKS Node Groups with Zero Downtime"
+authors: ["mike-metral"]
+tags: ["Kubernetes","EKS", "Migrations", "Node Groups"]
+date: "2019-07-22"
```
Comment (aydrian) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
date: "2019-07-23"
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md [hunk 44]
```diff
@@ -0,0 +1,174 @@
+---
+title: "Day 2 Kubernetes: Migrating EKS Node Groups with Zero Downtime"
+authors: ["mike-metral"]
+tags: ["Kubernetes","EKS", "Migrations", "Node Groups"]
```
Comment (aydrian) [inline content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]:
```suggestion
tags: ["Kubernetes","EKS"]
```

File: content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md (hunk 43 above)
Comment (metral) [reply content/blog/day-2-kubernetes-migrating-eks-nodegroups-with-zero-downtime/index.md]: Oops, forgot to save. Done and changes pushed.
